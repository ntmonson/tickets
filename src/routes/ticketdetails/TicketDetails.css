@import '../../components/variables.css';
@import '../../routes/addtickets/AddTickets.css';

.root {
  padding-left: 20px;
  padding-right: 20px;
}

.container {
  margin: 0 auto;
  padding: 0 0 40px;
  max-width: var(--screen-lg-min);
}

.ticketDetail {
  display: inline-block;
  padding: 10px 15px;
}

.detailHeader {
  text-decoration: underline;
  font-weight: bold;
  font-size: 1em;
}

.buttonBlue {
  background-color: #008cba;
  border: none;
  color: white;
  padding: 5px 24px;
  text-align: center;
  text-decoration: none;
  font-size: 16px;
  float: right;
  cursor: pointer;
}

.buttonClear {
  background-color: transparent;
  border: none;
  color: #00e;
  padding: 5px 24px;
  text-align: center;
  text-decoration: none;
  font-size: 16px;
  float: right;
  cursor: pointer;
}

button {
  background-color: Transparent;
  background-repeat: no-repeat;
  border: none;
  cursor: pointer;
  overflow: hidden;
  outline: none;
}

button:disabled,
button[disabled] {
  background-color: rgb(3, 53, 70);
  color: #666;
  cursor: not-allowed;
}
