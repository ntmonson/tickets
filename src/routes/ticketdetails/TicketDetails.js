import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { graphql, compose } from 'react-apollo';
import withStyles from 'isomorphic-style-loader/lib/withStyles';
import createPost from './createPost.graphql';
import getTicket from './getTicket.graphql';
import s from './TicketDetails.css';
import Posts from '../../components/Posts/Posts';

class TicketDetails extends Component {
  constructor(props) {
    super(props);
    this.state = {
      newTicketPost: '',
    };
  }
  handlePostChange = e => {
    this.setState({ newTicketPost: e.target.value });
  };

  handlePostClick = async () => {
    const { newTicketPost } = this.state;
    const { ticketId } = this.props;
    await this.props.addPost(newTicketPost, ticketId);
    this.clearInputState();
  };

  clearInputState = () => {
    this.setState({ newTicketPost: '' });
  };

  render() {
    if (this.props.data.loading) {
      return <div>Loading...</div>;
    }
    const ticket = this.props.data.databaseGetTicket;
    const lastPost = ticket.posts[ticket.posts.length - 1];
    return (
      <div className={s.root}>
        <div className={s.container}>
          <div className={s.ticketDetail}>
            <span className={s.detailHeader}>Ticket ID:</span>
            <br />
            {ticket.id}
          </div>
          <div className={s.ticketDetail}>
            <span className={s.detailHeader}>Topic:</span>
            <br />
            {ticket.topic}
          </div>
          {lastPost ? (
            <div className={s.ticketDetail}>
              <span className={s.detailHeader}>Last Post:</span>
              <br />
              {lastPost.createdAt}
            </div>
          ) : (
            <div className={s.ticketDetail}>
              <span className={s.detailHeader}>Last Post:</span>
              <br />
              No posts made yet.
            </div>
          )}
          <div className={s.ticketDetail}>
            <span className={s.detailHeader}>Status:</span>
            <br />
            {ticket.closed ? 'Closed' : 'Open'}
          </div>
          {ticket.posts.length ? <Posts posts={ticket.posts} /> : null}
          <div>
            <label htmlFor="post">
              Post
              <textarea
                type="textarea"
                maxLength="255"
                name="post"
                value={this.state.newTicketPost}
                onChange={this.handlePostChange}
              />
            </label>
          </div>
          <button
            disabled={!this.state.newTicketPost.length}
            className={s.buttonBlue}
            onClick={this.handlePostClick}
          >
            Submit
          </button>
          <button className={s.buttonClear} onClick={this.clearInputState}>
            Clear
          </button>
        </div>
      </div>
    );
  }
}

TicketDetails.propTypes = {
  ticketId: PropTypes.string.isRequired,
  data: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    databaseGetTicket: PropTypes.shape({
      closed: PropTypes.number.isRequired,
      createdAt: PropTypes.string.isRequired,
      id: PropTypes.string.isRequired,
      pinned: PropTypes.number.isRequired,
      posts: PropTypes.array.isRequired,
      topic: PropTypes.string.isRequired,
      updatedAt: PropTypes.string.isRequired,
    }),
  }).isRequired,
  addPost: PropTypes.func.isRequired,
};

const graphqlQueries = compose(
  graphql(getTicket, {
    options: ({ ticketId }) => ({ variables: { id: ticketId } }),
  }),
  graphql(createPost, {
    name: 'addPost',
    props: ({ addPost }) => ({
      addPost: (content, ticketId) =>
        addPost({ variables: { content, ticketId } }),
    }),
    options: {
      update: (proxy, { data: { databaseCreatePost } }) => {
        const data = proxy.readQuery({
          query: getTicket,
          variables: { id: databaseCreatePost.ticketId },
        });
        data.databaseGetTicket.posts.push(databaseCreatePost);
        proxy.writeQuery({ query: getTicket, data });
      },
    },
  }),
);
export default compose(withStyles(s), graphqlQueries)(TicketDetails);
