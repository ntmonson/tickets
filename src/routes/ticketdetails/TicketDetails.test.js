/* eslint-env jest */
/* eslint-disable padded-blocks, no-unused-expressions */

import React from 'react';
import renderer from 'react-test-renderer';
import '@babel/polyfill';
import { TicketDetails } from './TicketDetails';

describe('TicketDetails', () => {
  const data = {
    loading: false,
    databaseGetTicket: {
      closed: 1,
      createdAt: '1234',
      id: '1234',
      pinned: 1,
      posts: [],
      topic: 'Topic',
      updatedAt: '05/22/2018',
    },
  };
  test('TicketDetails renders correctly, loading false, closed 1, pinned 1', () => {
    const component = renderer
      .create(
        <TicketDetails ticketId="ASDF123" addPost={() => {}} data={data} />,
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  test('TicketDetails renders correctly, loading true, closed 1, pinned 1', () => {
    data.loading = true;
    const component = renderer
      .create(
        <TicketDetails ticketId="ASDF123" addPost={() => {}} data={data} />,
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  test('TicketDetails renders correctly, loading false, closed 0, pinned 1', () => {
    data.loading = false;
    data.closed = 0;
    const component = renderer
      .create(
        <TicketDetails ticketId="ASDF123" addPost={() => {}} data={data} />,
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  test('TicketDetails renders correctly, loading false, closed 0, pinned 0', () => {
    data.pinned = 0;
    const component = renderer
      .create(
        <TicketDetails ticketId="ASDF123" addPost={() => {}} data={data} />,
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
