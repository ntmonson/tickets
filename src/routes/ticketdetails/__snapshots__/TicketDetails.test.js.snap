// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketDetails TicketDetails renders correctly, loading false, closed 0, pinned 0 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <div
      className="ticketDetail"
    >
      <span
        className="detailHeader"
      >
        Ticket ID:
      </span>
      <br />
      1234
    </div>
    <div
      className="ticketDetail"
    >
      <span
        className="detailHeader"
      >
        Topic:
      </span>
      <br />
      Topic
    </div>
    <div
      className="ticketDetail"
    >
      <span
        className="detailHeader"
      >
        Last Post:
      </span>
      <br />
      No posts made yet.
    </div>
    <div
      className="ticketDetail"
    >
      <span
        className="detailHeader"
      >
        Status:
      </span>
      <br />
      Closed
      <button
        onClick={[Function]}
        style={
          Object {
            "color": "Red",
            "fontSize": "1.25em",
            "marginLeft": "5px",
          }
        }
      >
        <i
          className="fas fa-toggle-on"
        />
      </button>
    </div>
    <div>
      <label
        htmlFor="post"
      >
        Post
        <textarea
          maxLength="255"
          name="post"
          onChange={[Function]}
          type="textarea"
          value=""
        />
      </label>
    </div>
    <button
      className="buttonBlue"
      disabled={true}
      onClick={[Function]}
    >
      Submit
    </button>
    <button
      className="buttonClear"
      onClick={[Function]}
    >
      Clear
    </button>
  </div>
</div>
`;

exports[`TicketDetails TicketDetails renders correctly, loading false, closed 0, pinned 1 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <div
      className="ticketDetail"
    >
      <span
        className="detailHeader"
      >
        Ticket ID:
      </span>
      <br />
      1234
    </div>
    <div
      className="ticketDetail"
    >
      <span
        className="detailHeader"
      >
        Topic:
      </span>
      <br />
      Topic
    </div>
    <div
      className="ticketDetail"
    >
      <span
        className="detailHeader"
      >
        Last Post:
      </span>
      <br />
      No posts made yet.
    </div>
    <div
      className="ticketDetail"
    >
      <span
        className="detailHeader"
      >
        Status:
      </span>
      <br />
      Closed
      <button
        onClick={[Function]}
        style={
          Object {
            "color": "Red",
            "fontSize": "1.25em",
            "marginLeft": "5px",
          }
        }
      >
        <i
          className="fas fa-toggle-on"
        />
      </button>
    </div>
    <div>
      <label
        htmlFor="post"
      >
        Post
        <textarea
          maxLength="255"
          name="post"
          onChange={[Function]}
          type="textarea"
          value=""
        />
      </label>
    </div>
    <button
      className="buttonBlue"
      disabled={true}
      onClick={[Function]}
    >
      Submit
    </button>
    <button
      className="buttonClear"
      onClick={[Function]}
    >
      Clear
    </button>
  </div>
</div>
`;

exports[`TicketDetails TicketDetails renders correctly, loading false, closed 1, pinned 1 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <div
      className="ticketDetail"
    >
      <span
        className="detailHeader"
      >
        Ticket ID:
      </span>
      <br />
      1234
    </div>
    <div
      className="ticketDetail"
    >
      <span
        className="detailHeader"
      >
        Topic:
      </span>
      <br />
      Topic
    </div>
    <div
      className="ticketDetail"
    >
      <span
        className="detailHeader"
      >
        Last Post:
      </span>
      <br />
      No posts made yet.
    </div>
    <div
      className="ticketDetail"
    >
      <span
        className="detailHeader"
      >
        Status:
      </span>
      <br />
      Closed
      <button
        onClick={[Function]}
        style={
          Object {
            "color": "Red",
            "fontSize": "1.25em",
            "marginLeft": "5px",
          }
        }
      >
        <i
          className="fas fa-toggle-on"
        />
      </button>
    </div>
    <div>
      <label
        htmlFor="post"
      >
        Post
        <textarea
          maxLength="255"
          name="post"
          onChange={[Function]}
          type="textarea"
          value=""
        />
      </label>
    </div>
    <button
      className="buttonBlue"
      disabled={true}
      onClick={[Function]}
    >
      Submit
    </button>
    <button
      className="buttonClear"
      onClick={[Function]}
    >
      Clear
    </button>
  </div>
</div>
`;

exports[`TicketDetails TicketDetails renders correctly, loading true, closed 1, pinned 1 1`] = `
<div>
  Loading...
</div>
`;
